RPI/src/processor.py

```diff
@@ -33,58 +33,139 @@ def __init__(self, stringer, protocol_handler, expectation_handler):
+                pass    # TODO implement
+            elif input_ == "Black Set":
+                pass    # TODO implement
+
+            # Error interactions:
+            elif input_ == "Error Occurred":
+                pass    # TODO implement
+            else:
+                return ["Unknown Message"]
+        except Exception:
+            # TODO add error handling code
+            pass
 
     def __ping(self):
         """Returns output in case of ping."""
 
         self.__expectation_handler.ping()
+        self.__protocol_handler.inform_alive()
         expired_outputs = self.__expectation_handler.get_expired_outputs()
-
-        if expired_outputs:
-            return expired_outputs
-        else:
-            return [b"Pong"]
+        return ["Pong"] + expired_outputs
 
     def __motion_detected(self):
         """Returns output in case of motion."""
 
         if not self.__stringer.is_complete():
             # The stringer still needs disks
-            self.__expectation_handler.add(b"Confirm Extend Blocker", [b"Retract Blocker"], 10)
-            return [b"Extend Blocker", b"Scan Color"]
+            self.__expectation_handler.add("Confirm Blocker Extended",
+                                           ["Retract Blocker"], 10)    # TODO adjust timer
+            return ["Extend Blocker"]
         else:
             # The stringer does not need disks
-            return [b"Ignore"]
+            return ["Ignore"]
 
-    def __color_detected(self, color):
+    def __primary_color_detected(self, color):
         """
-        Returns output in case of color.
+        Removes expectation and returns output in case of color.
         @param color  The color that was detected.
         """
 
+        self.__expectation_handler.remove("Primary Color Detected")
+
         if not self.__stringer.should_pickup(color):
             # We do not want the color
-            self.__expectation_handler.add(b"Confirm Retract Blocker", [b"Extend Blocker"], 10)
-            return [b"Retract Blocker"]
+            self.__expectation_handler.add("Confirm Blocker Retracted",
+                                           ["Extend Blocker"], 10)     # TODO adjust timer
+            return ["Retract Blocker"]
         elif not self.__protocol_handler.can_pickup():
             # We are not allowed to pick up a disk
-            return [b"Retract Blocker"]
+            self.__expectation_handler.add("Confirm Blocker Retracted",
+                                           ["Extend Blocker"], 10)     # TODO adjust timer
+            return ["Retract Blocker"]
         else:
             # We want the color and we are allowed to pick up a disk
-            return [b"Push", b"Retract Blocker"]
+            self.__expectation_handler.add("Confirm Pusher Push", [], 10)                          # TODO adjust timer
+            self.__expectation_handler.add("Secondary Color Detected " + str(color), [], 10)     # TODO adjust timer
```

Comment:
Secondary Color Detected + str(color) is not an existing input, I think this would have to be Secondary White or Secondary Black.